### PR TITLE
Add AppleII UVX shape to Block class and magnetization argument to Block radia creation method

### PR DESCRIPTION
**Pull request by xresende**

### CODE REVIEW

**Comment by lucashfrancisco** [comment]
A note on the magnetization vector.

When the Block object is initialized, the material is set as the input material (if not None) or as a default material that uses the magnetization for determining remanence modulus.
When the Radia object is created, it recieves both the Material and the magnetization modulus. But the magnetization modulus determines only the easy magnetization axis. I think adding the further option of changing magnetization when manually calling the create_radia_object method would create some confusion.

block_material drawio

**Review by lucashfrancisco — COMMENTED** [review]
A note on the magnetization vector.

When the Block object is initialized, the material is set as the input material (if not None) or as a default material that uses the magnetization for determining remanence modulus.
When the Radia object is created, it recieves both the Material and the magnetization modulus. But the magnetization modulus determines only the easy magnetization axis. Radia documentation says:

"""
The magnetization of the block is initialized to {mx,my,mz}. If any magnetic material is later applied to the block, it fully overrides this value of the magnetization. At the relaxation, the value of the magnetization will be further modified.
"""

I think adding the further option of changing magnetization when manually calling the create_radia_object method would create some confusion. The magnetization set at create_radia_object would alter only magnetization direction, and never its modulus. The magnetization set on __init__ would set both remanence and magnetization direction (if a material is not provided).

block_material drawio

**Review by lucashfrancisco — COMMENTED** [review]
Mistakenly self-requested review. Comments are on previous review.

**Comment by xresende** [comment]
> A note on the magnetization vector.
> 
> When the Block object is initialized, the material is set as the input material (if not None) or as a default material that uses the magnetization for determining remanence modulus. When the Radia object is created, it recieves both the Material and the magnetization modulus. But the magnetization modulus determines only the easy magnetization axis. Radia documentation says:
> 
> At object creation function: "The magnetization of the block is initialized to {mx,my,mz}. If any magnetic material is later applied to the block, it fully overrides this value of the magnetization. At the relaxation, the value of the magnetization will be further modified."
> 
> At linear material creation function: "(..) the orientation of the remanent magnetization vector (giving the orientation of the easy magnetization axis) is specified by the initial magnetization vector in the object to which the material is later applied. If mr>0 (mr<0), the remanent magnetization is set to be parallel (antiparallel) to the initial magnetization vector in the object. The magnitude of the remanent magnetization vector is |mr|."
> 
> I think adding the further option of changing magnetization when manually calling the create_radia_object method would create some confusion. The magnetization set at create_radia_object would alter only magnetization direction, and never its modulus. The magnetization set on **init** would set both remanence and magnetization direction (if a material is not provided).
> 
> block_material drawio

@lucashfrancisco , I understand that implementations of imaids and radia use the initial magnetization vector to define the easy direction. but they also set the H=0 magnetization of the block (M0) using the attribute self._magnetization. otherwise, at what arbitrary direction would M0 be pointed to? by adding the `magnetization` argument to `create_radia_object` we are simply repeating what is already done in the class constructor, as it has `magnetization` as argument as well. am i still missing something ?

and we did try this new  `create_radia_object` with a mag argument and the fields generated by the block are changed. we tried a mag vector with a rather large modulus and its calculated field was correspondingly large. I think @Gabrielrezende-asc will prepare a code snippet showing this.

what i think is missing in this PR is updating `self._material` based on the mag vector passed to `create_radia_object`. do you agree ?

the reason i added this method argument is because we did not find a natural and easy way to alter the mag vector of a block that belongs to a cassette/insertion_device (we need this in order to implement model calibration algorithms). maybe we are just not familiar with the package yet. how would you do this ?

**Comment by Gabrielrezende-asc** [comment]
> A note on the magnetization vector.
> 
> When the Block object is initialized, the material is set as the input material (if not None) or as a default material that uses the magnetization for determining remanence modulus. When the Radia object is created, it recieves both the Material and the magnetization modulus. But the magnetization modulus determines only the easy magnetization axis. Radia documentation says:
> 
> At object creation function: "The magnetization of the block is initialized to {mx,my,mz}. If any magnetic material is later applied to the block, it fully overrides this value of the magnetization. At the relaxation, the value of the magnetization will be further modified."
> 
> At linear material creation function: "(..) the orientation of the remanent magnetization vector (giving the orientation of the easy magnetization axis) is specified by the initial magnetization vector in the object to which the material is later applied. If mr>0 (mr<0), the remanent magnetization is set to be parallel (antiparallel) to the initial magnetization vector in the object. The magnitude of the remanent magnetization vector is |mr|."
> 
> I think adding the further option of changing magnetization when manually calling the create_radia_object method would create some confusion. The magnetization set at create_radia_object would alter only magnetization direction, and never its modulus. The magnetization set on **init** would set both remanence and magnetization direction (if a material is not provided).
> 
> block_material drawio

Hi Lucas, as Ximenes said, we made some tests changing the magnetization vector of a block. We verified that when we change the magnitude of magnetization using create_radia_object, the field generate by this block also changes. Here we have a simple script showing this:
```block_test = epu.cassettes['csd'].blocks[50]
print("Default magnetization: ",block_test.magnetization)
field_before = block_test.get_field_at_point(point=[0,0,0])
print("Original field: ",field_before)
print("Field intensity: {:.3e}".format(np.sqrt(np.sum(np.square(field_before)))))

block_test.create_radia_object(magnetization =[0.0, 5.0, 0.0])
print("New magnetizarion: ",block_test.magnetization)
field_after = block_test.get_field_at_point(point=[0,0,0])
print("New field: ",field_after)
print("Field intensity: {:.3e}".format(np.sqrt(np.sum(np.square(field_after)))))```

**Comment by lucashfrancisco** [comment]
@Gabrielrezende-asc, thanks for the snippet. Indeed the field changes.
I find it courious that it does work and the field does change, since the Radia documentation seems to say otherwise, as I've understood it, at least.

I thought that maybe the material object itself was being altered automatically, but it is not, since if i do:

```python
block_shape = [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
block_thickness = 12.375
z_position = 0
magnetization = [0, 1.24, 0]

epu_block = Block(shape=block_shape,
                  length=block_thickness,
                  subdivision=[[1, 1, 1]],
                  longitudinal_position=0,
                  magnetization=magnetization)

print('field before', epu_block.get_field_at_point([0,2,0]))
print('mr before', epu_block._material.mr)
epu_block.create_radia_object(magnetization=[0,10,0])
print('field after', epu_block.get_field_at_point([0,2,0]))
print('mr after', epu_block._material.mr)
```

I get:
```
field before [-0.3260840044222332, 0.21836083517215366, 3.0424574709591263e-12]
mr before 1.24
field after [-2.629709712972459, 1.7609744773469338, -3.1552427671235395e-11]
mr after 1.24
```

So the material stays the same and the field changes. This behaviour is strange to me. Maybe Radia documentation is outdated in relation to the material/magnetization behavior?

**Comment by lucashfrancisco** [comment]
Either way, @xresende. I agree that the issue I pointed out is basically the need for keeping `self._material` up to date.  But it is indeed it is inconvenient to do it this way (in the create_radia_object method). It seems that we are doing something kind of circular with the material definition.

I think that instead of a method for copying the block, a more natural way would be a setter method that changes the magnetization, recreates material and recreates the radia object. What do you think?

Basically "re-do with other magnetization" when we give a block.magnetization = [mx, my, mz] command.

**Comment by xresende** [comment]
> Either way, @xresende. I agree that the issue I pointed out is basically the need for keeping `self._material` up to date. But it is indeed it is inconvenient to do it this way (in the create_radia_object method). It seems that we are doing something kind of circular with the material definition.
> 
> I think that instead of a method for copying the block, a more natural way would be a setter method that changes the magnetization, recreates material and recreates the radia object. What do you think?
> 
> Basically "re-do with other magnetization" when we give a block.magnetization = [mx, my, mz] command.

we can implement a magnetization setter. But it will basically replicate what create_radia_object does. but the class interface will be cleaner this way. I'll git it a try...

**Comment by lucashfrancisco** [comment]
Agree, it would indeed be the same as create_radia_object in the Block case. Still, it might be a useful abstraction to have a create_radia_object method, that directly creates the object, and a separate setter that happens to call the radia object creation because it is necessary to do so when the magnetization is changed.

**Comment by xresende** [comment]
I reverted `create_radia_object` method and implemented the block magnetization setter. I noticed that the the application of material uses the `subblock`object and its magnetization for easy axis determination and thus no further material setting is needed other than updating the mag vector.

**Comment by lucashfrancisco** [comment]
@xresende, I added one final touch to the magnetization setter (updating the material) and I think the branch is now ok to merge.

I made a test for:
```python
block_shape = [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
block_thickness = 12.375
z_position = 0
magnetization = [0, 1.24, 0]

epu_block = Block(shape=block_shape, length=block_thickness,
                  subdivision=[[1, 1, 1]], longitudinal_position=0,                  
                  magnetization=magnetization, name='EPU Block',
                  ksipar=0.3, ksiper=0.2)
```
By doing the field calculation and updating the magnetization:
```python
print('\nfield before: ', epu_block.get_field_at_point([0,2,0]), '\\\\\n')
print(epu_block)

epu_block.magnetization = [0, 10.0, 0]

print('\nfield after: ', epu_block.get_field_at_point([0,2,0]), '\n')
print(epu_block)
```
The result was ok, everyting updated:
```
field before:  [-0.3260840044081152, 0.21836083515945737, -9.506248568012464e-12] \\

shape              : [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
length             : 12.375
magnetization      : [0, 1.24, 0]
subdivision        : [[1, 1, 1]]
rectangular        : False
longitudinal_position : 0
use_default_material : True
default_material_kwargs : {'ksipar': 0.3, 'ksiper': 0.2}
material           : linear             : True
mr                 : 1.24
ksipar             : 0.3
ksiper             : 0.2
hlist              : None
mlist              : None
name               : 
radia_object       : 4474

name               : EPU Block
radia_object       : 4477


field after:  [-2.6297097132162275, 1.760974477401688, 2.5437811296121778e-11] 

shape              : [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
length             : 12.375
magnetization      : [0, 10.0, 0]
subdivision        : [[1, 1, 1]]
rectangular        : False
longitudinal_position : 0
use_default_material : True
default_material_kwargs : {'ksipar': 0.3, 'ksiper': 0.2}
material           : linear             : True
mr                 : 10.0
ksipar             : 0.3
ksiper             : 0.2
hlist              : None
mlist              : None
name               : 
radia_object       : 4478

name               : EPU Block
radia_object       : 4481
```

As a side comment, the issue of "lacking garbage collector" you mentioned in other instance is seen here by the raida objects int identifier.

**Review by lucashfrancisco — APPROVED** [review]
@xresende, I added one final touch to the magnetization setter (updating the material) and I think the branch is now ok to merge.

I made a test for:
```python
block_shape = [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
block_thickness = 12.375
z_position = 0
magnetization = [0, 1.24, 0]

epu_block = Block(shape=block_shape, length=block_thickness,
                  subdivision=[[1, 1, 1]], longitudinal_position=0,                  
                  magnetization=magnetization, name='EPU Block',
                  ksipar=0.3, ksiper=0.2)
```
By doing the field calculation and updating the magnetization:
```python
print('\nfield before: ', epu_block.get_field_at_point([0,2,0]), '\\\\\n')
print(epu_block)

epu_block.magnetization = [0, 10.0, 0]

print('\nfield after: ', epu_block.get_field_at_point([0,2,0]), '\n')
print(epu_block)
```
The result was ok, everyting updated:
```
field before:  [-0.3260840044081152, 0.21836083515945737, -9.506248568012464e-12] \\

shape              : [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
length             : 12.375
magnetization      : [0, 1.24, 0]
subdivision        : [[1, 1, 1]]
rectangular        : False
longitudinal_position : 0
use_default_material : True
default_material_kwargs : {'ksipar': 0.3, 'ksiper': 0.2}
material           : linear             : True
mr                 : 1.24
ksipar             : 0.3
ksiper             : 0.2
hlist              : None
mlist              : None
name               : 
radia_object       : 4474

name               : EPU Block
radia_object       : 4477


field after:  [-2.6297097132162275, 1.760974477401688, 2.5437811296121778e-11] 

shape              : [[[0.1, 0], [40, 0], [40, -40], [0.1, -40]]]
length             : 12.375
magnetization      : [0, 10.0, 0]
subdivision        : [[1, 1, 1]]
rectangular        : False
longitudinal_position : 0
use_default_material : True
default_material_kwargs : {'ksipar': 0.3, 'ksiper': 0.2}
material           : linear             : True
mr                 : 10.0
ksipar             : 0.3
ksiper             : 0.2
hlist              : None
mlist              : None
name               : 
radia_object       : 4478

name               : EPU Block
radia_object       : 4481
```

As a side comment, the issue of "lacking garbage collector" you mentioned in other instance is seen here by the raida objects int identifier.

File: imaids/blocks.py [hunk 1]
```diff
@@ -211,6 +214,11 @@ def magnetization(self, new_magnetization):
         """Set new block magnetization vector [T]."""
         # update magnetization attribute
         self._magnetization = self._check_magnetization(new_magnetization)
+        # replace material object.
+        if self._use_default_material:
```
Comment (xresende) [inline imaids/blocks.py:218]: @lucashfrancisco , i did not include material updating because create_radia_object already uses the subblock mag vector for easy axis and the material type is already set (from __init__) when mag setter is invoked. i dont think material updating is necessary, despite what we talked last week.

Comment (lucashfrancisco) [reply imaids/blocks.py:218]: It is probably not necessary for our use. But I believe it would be a good consistency practice to keep all Radia objects up to date, including the material Radia object referenced by the block (referenced by the imaids.Material object which is referenced by the block).
Also, this way it will not happen that one prints the object, like I did above, and the mr value shown in the material section does not correspond to the actual block remanence.

Maybe this shows an issue that the "compicated" mixed way in which the magnetization is set (vector + material  arguments) may not be ideal. But i guess this is a separate issue.

File: imaids/blocks.py [hunk 2]
```diff
@@ -349,3 +362,8 @@ def bounding_box_center(self):
         center = _np.mean(self.bounding_box(), axis=1)
         
         return list(center)
+
+    def _check_magnetization(self, magnetization):
+        if len(magnetization) != 3:
+            raise ValueError('Invalid magnetization argument.')
+        return magnetization
```
Comment (fernandohds564) [inline imaids/blocks.py:369]: there is no need to return magnetization here, only a check would suffice. Maybe we could consider returning `True` or `False` instead of `ValueError`.

**Review by fernandohds564 — APPROVED** [review]

**Review by Gabrielrezende-asc — APPROVED** [review]

**Review by FernandaMenicucci — APPROVED** [review]

**Comment by lucashfrancisco** [comment]
@xresende, may this branch be merged?

File: imaids/blocks.py (hunk 2 above)
Comment (fernandohds564) [reply imaids/blocks.py:348]: hi @lucashfrancisco , what do you think about this?

Comment (xresende) [reply imaids/blocks.py:348]: @fernando, I return magnetization here so that __init__ could both check its value and receive it after checking before settings its _magnetization attribute.

**Comment by xresende** [comment]
> @xresende, may this branch be merged?

for me it can. just waiting for @fernandohds564 on the issue he raise about the implementation of  `_check_magnetization`...

File: imaids/blocks.py (hunk 1 above)
Comment (xresende) [reply imaids/blocks.py:344]: I too found it not natural how material/magnetization is set...

File: imaids/blocks.py (hunk 2 above)
Comment (fernandohds564) [reply imaids/blocks.py:474]: But why is that useful? Wouldn't be more pythonic to not return anything or return a boolean indicating if the value is ok?

Comment (xresende) [reply imaids/blocks.py:474]: it is useful because this checking has to be done in `__init__` and also in `magnetization` setter. By returning the magnetization we can condense the vector checking and the attribute `_magnetization` in one line. Returning a boolean and processing it in these two methods effectively is replicating code which could be only in `_check_magnetization` into the two calling functions. For example:

```
__init__:
     status =  self._check_magnetization(magnetization)
     if not status:
         raise ValueError('Invalid magnetization argument.')
      self._magnetization = self._check_magnetization(magnetization)
  
@magnetization.setter:
     status =  self._check_magnetization(magnetization)
     if not status:
         raise ValueError('Invalid magnetization argument.')
     self._magnetization = self._check_magentization(magnetization)
```

compared to 

```
__init__:
      self._magnetization = self._check_magentization(magnetization)
  
@magnetization.setter:
     self._magnetization = self._check_magentization(magnetization)
```

Comment (fernandohds564) [reply imaids/blocks.py:474]: Well... I disagree with your first implementation... actually it would be something like this:
```
__init__:
      self.magnetization = magnetization
  
@magnetization.setter:
     if len(magnetization) != 3:
         raise ValueError('Invalid magnetization argument.')
     self._magnetization = magnetization
```
It seems that even the existence of the method `_check_magnetization` is not needed in this case and we would have the advantage of making use of the setter appropriately...
What do you think?

Comment (xresende) [reply imaids/blocks.py:474]: oh, I see, @fernandohds564 . Now, I understand your proposal. I did not see your last message before talking to you in person moments agor. I agree. I will do it.

**Comment by xresende** [comment]
> @xresende, may this branch be merged?

@lucashfrancisco , now I sorted things out with @fernandohds564 . I implemented his suggestions and for me you can finally merge this PR. sorry, this one took unusual time to converge, partially due to my vacations! :-)
